### PR TITLE
[TP] Add wildcard support for *

### PR DESCRIPTION
**Pull request by kwen2501**
This PR adds support for the following specification:
```
parallelize_module(
    model,
    device_mesh,
    {"layers.*.linear" : ColwiseParallel},
)
``` 
where `*` replaces what would typically be 0, 1, 2, ... (layer ids).

Rules:
- `*` can be at any hierarchy of the path;
- Multiple `*`s are supported, e.g. `*.*.linear`;
- Mixing `*` with other characters are not supported at the moment, e.g. `layer*.linear` (though it may be useful too).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122966
* __->__ #122923
* #122919

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

**Review by wanchaol — COMMENTED** [review]
looks like a good UX improvements! Please add tests.

**Comment by kwen2501** [comment]
Thanks, will add in a next PR.

File: torch/distributed/tensor/parallel/api.py [hunk 1]
```diff
@@ -88,17 +88,38 @@ def parallelize_module(  # type: ignore[return]
...
-            parent_module = module
-            if "." in module_path:
-                path_splits = module_path.split(".")
-                parent_module_path = ".".join(path_splits[:-1])
-                parent_module = module.get_submodule(parent_module_path)
-                module_path = path_splits[-1]
+            parent_module = leaf_module = module
+            path_splits = module_path.split(".")
+            if len(path_splits) == 0:
+                raise RuntimeError(
```
[the remark this answers is not in the document]
Comment (kwen2501) [reply torch/distributed/tensor/parallel/api.py:94]: Thanks! Patched in #122966

**Comment by kwen2501** [comment]
Abandoning in favor of #122968 which adds wildcard support as a whole.

**Pull request closed without merging**